src/flow_solver/flow_solver_cases/periodic_entropy_tests.cpp

```diff
@@ -0,0 +1,290 @@
+        metric_oper.build_volume_metric_operators(
+            n_quad_pts, n_grid_nodes,
+            mapping_support_points,
+            mapping_basis,
+            dg->all_parameters->use_invariant_curl_form);
+
+        // Fetch the modal soln coefficients
+        // We immediately separate them by state as to be able to use sum-factorization
+        // in the interpolation operator. If we left it by n_dofs_cell, then the matrix-vector
+        // mult would sum the states at the quadrature point.
+        // That is why the basis functions are based off the 1state oneD fe_collection.
+        std::array<std::vector<double>,nstate> soln_coeff;
+        for (unsigned int idof = 0; idof < n_dofs_cell; ++idof) {
+            const unsigned int istate = dg->fe_collection[poly_degree].system_to_component_index(idof).first;
+            const unsigned int ishape = dg->fe_collection[poly_degree].system_to_component_index(idof).second;
+            if(ishape == 0){
+                soln_coeff[istate].resize(n_shape_fns);
+            }
+            soln_coeff[istate][ishape] = dg->solution(dofs_indices[idof]);
+        }
+        // Interpolate each state to the quadrature points using sum-factorization
+        // with the basis functions in each reference direction.
+        std::array<std::vector<double>,nstate> soln_at_q;
+        for(int istate=0; istate<nstate; istate++){
+            soln_at_q[istate].resize(n_quad_pts);
+            // Interpolate soln coeff to volume cubature nodes.
+            soln_basis.matrix_vector_mult_1D(soln_coeff[istate], soln_at_q[istate],
+                                             soln_basis.oneD_vol_operator);
+        }
+
+        // Loop over quadrature nodes, compute quantities to be integrated, and integrate them.
+        for (unsigned int iquad=0; iquad<n_quad_pts; ++iquad) {
+
+            std::array<double,nstate> soln_state;
+            // Extract solution and gradient in a way that the physics ca n use them.
+            for(int istate=0; istate<nstate; istate++){
+                soln_state[istate] = soln_at_q[istate][iquad];
+            }
+            const double density = soln_state[0];                                                                                                                                                              
+            const double pressure = euler_physics->compute_pressure(soln_state);
+            const double entropy = log(pressure) - 1.4* log(density);
+            const double quadrature_entropy = -density*entropy/0.4;
+            //integrand_numerical_entropy_function = this->euler_physics->compute_numerical_entropy_function(soln_state); //SOMEHOW NOT REACHING PHYSICS
+            integrand_numerical_entropy_function = quadrature_entropy;
+            integral_numerical_entropy_function += integrand_numerical_entropy_function * quad_weights[iquad] * metric_oper.det_Jac_vol[iquad];
```

Comment:
This form integrates across the solution nodes rather than using an inner product.

From our discussions in the fall semester, using `num-entropy = v M u` was not generally true for Euler FR. We may be able to use if for cDG, but I'm not sure if it is proven or robust.